src/Orleans/Streams/Internal/StreamHandshakeToken.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+Project Orleans Cloud Service SDK ver. 1.0
+ 
+Copyright (c) Microsoft Corporation
+ 
+All rights reserved.
+ 
+MIT License
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
+associated documentation files (the ""Software""), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+using System;
+
+namespace Orleans.Streams
+{
+    [Serializable]
+    public class StreamHandshakeToken : IEquatable<StreamHandshakeToken>
+    {
+        public StreamSequenceToken Token { get; private set; }
+
+        [Serializable] private class StartToken : StreamHandshakeToken { }
+        [Serializable] private class DeliveryToken : StreamHandshakeToken { }
+
+        public static StreamHandshakeToken CreateStartToken(StreamSequenceToken token)
+        {
+            if (token == null) return default(StreamHandshakeToken);
+            return new StartToken { Token = token };
+        }
+
+        public static StreamHandshakeToken CreateDeliveyToken(StreamSequenceToken token)
+        {
+            return new DeliveryToken { Token = token };
+        }
+
+        public bool Equals(StreamHandshakeToken other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+            if (other.GetType() != GetType()) return false;
```

Comment:
Internal types serialized correctly, but I don't know if serializers were generated so I moved them out to be safe.